pkg/controller/bluegreendeployment/controller.go

```diff
@@ -0,0 +1,195 @@
+func (c *BlueGreenDeploymentController) sync(b *controllerv1alpha1.BlueGreenDeployment) error {
+	var err error
+
+	// Initialize .status.activeReplicaSetColor field to blue color if it is not set yet
+	if b.Status.ActiveReplicaSetColor == "" {
+		// Update .status.activeReplicaSetColor field to blue color at the beginning
+		if b, err = updateBlueGreenDeployment(c.blueGreenDeploymentClient, b.Name, b.Namespace, func(bgd *controllerv1alpha1.BlueGreenDeployment) {
+			bgd.Status.ActiveReplicaSetColor = BlueColor
+		}); err != nil {
+			return fmt.Errorf("failed to update .status.activeReplicaSetColor field of BlueGreenDeployment: %v", err)
+		}
+	}
+
+	blueRS, greenRS, err := createReplicaSetsIfNotFound(b, c.k8sClientSet)
+	if err != nil {
+		return err
+	}
+
+	service, err := createServiceIfNotFound(b, c.k8sClientSet)
+	if err != nil {
+		return err
+	}
+
+	err = c.reconcileActiveReplicaSet(b, blueRS, greenRS, service)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// reconcileActiveReplicaSet checks active and inactive ReplicaSets' pod specs to
+// see if any of the pod specs matches BlueGreenDeployment object's pod spec.
+// If the active ReplicaSet has the matching spec, the controller does nothing.
+// Else if the inactive ReplicaSet has the matching spec, the controller:
+// 1. scales up the inactive ReplicaSet
+// 2. modifies label selectors of the service to point to the newly active ReplicaSet
+// 3. scales down previously active ReplicaSet to make it inactive
+// Else (i.e., none of the Replicasets has the matching spec), the controller:
+// 1. deletes the inactive ReplicaSet to make room for a new ReplicaSet
+// 2. creates and scales up the new ReplicaSet
+// 3. modifies label selectors of the service to point to the new ReplicaSet
+// 4. scales down previously active ReplicaSet to make it inactive
+// The function updates .status.activeReplicaSetColor field and the service's label
+// selectors with color of the active ReplicaSet.
+func (c *BlueGreenDeploymentController) reconcileActiveReplicaSet(b *controllerv1alpha1.BlueGreenDeployment, blueRS, greenRS *appsv1.ReplicaSet, service *corev1.Service) error {
+	var err error
+
+	// If a ReplicaSet has the same color as indicated by .status.activeReplicaSetColor field
+	// of BlueGreenDeployment object, it is the active ReplicaSet.
+	activeColor := b.Status.ActiveReplicaSetColor
+	activeRS := blueRS
+	inactiveRS := greenRS
+	if activeColor == GreenColor {
+		activeRS = greenRS
+		inactiveRS = blueRS
+	}
+
+	// Hash pod spec of active ReplicaSet, inactive ReplicaSet, and BlueGreenDeployment object for comparison
+	hashedActiveRSPodSpec := bgdutil.ComputeHash(&activeRS.Spec.Template.Spec)
+	hashedInactiveRSPodSpec := bgdutil.ComputeHash(&inactiveRS.Spec.Template.Spec)
+	hashedBGDPodSpec := bgdutil.ComputeHash(&b.Spec.PodSpec)
+
+	// Case: active ReplicaSet has the matching spec
+	if reflect.DeepEqual(hashedActiveRSPodSpec, hashedBGDPodSpec) {
```

Comment:
Don't compare hash, because:

1. Hash collisions
2. Any API change, such as adding new fields would change the hash value

Just compare the spec.

Reply:
I was originally comparing the pod specs, but I was then suggested to use hashing.

For hash collisions, we can add a field named `CollisionCount` to prevent them. I didn't include the field to make the API simple.

For API change, I don't think it matters - the controller computes the hash during every sync loop. Correct me if I am wrong.

Reply:
Imagine that spec A and B generates the same hash value X because of hash collisions. You'd think they are the same when comparing, but they're actually different. 

`CollisionCount` isn't used for this purpose.

Reply:
Is there any reason why `CollisionCount` does not work? I referred to the this [code snippet](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/controller_utils.go#L1031)

Only difference is it computes hash for `PodTemplateSpec`, not `PodSpec`. Either way, it should be able to avoid hash collision.

Reply:
`CollisionCount` is to solve the problem of name collisions (due to hash collisions). Hash is not used for comparison in any workload controllers, but for generating unique labels/names.

Reply:
Convinced. `CollisionCount` should solve name collisions, not pod spec collisions. Furthrmore, pod spec comparison is slightly faster. :)

Reply:
@pwittrock I confirmed with @janetkuo that all workload controllers now compare pod spec directly. There was a change from hash comparison to pod spec comparison for the controllers before due to hash collision. I think we should revert back to on-the-fly jsonified pod spec comparison.